File: next_permutation.cpp
/*The next permutation of an array of integers is the next lexicographically greater permutation of its integer. More formally, if all the permutations of the array are sorted in one container according to their lexicographical order, then the next permutation of that array is the permutation that follows it in the sorted container. If such arrangement is not possible, the array must be
rearranged as the lowest possible order (i.e., sorted in ascending order).*/
#include <bits/stdc++.h>
#define db1(x) cerr << #x << "=" << x << '\n'
#define db2(x, y) cerr << #x << "=" << x << "," << #y << "=" << y << '\n'
#define db3(x, y, z) cerr << #x << "=" << x << "," << #y << "=" << y << "," << #z << "=" << z << '\n'
#define rep(i, n) for (ll i = 0; i < (n); ++i)
#define repA(i, a, n) for (ll i = a; i <= (n); ++i)
#define repD(i, a, n) for (ll i = a; i >= (n); --i)
#define mp make_pair
#define sp << ' ' <<
#define pb push_back
#define FastIO               \
    ios::sync_with_stdio(0); \
    cin.tie(0);              \
    cout.tie(0);
using namespace std;
using ll = long long;
void next_permutation(vector<int> &A)
{
    // vector<int> res;
    //  reverse travel to find next greater
    int n = A.size();
    int pi = 0, pj = 0;
    int desc = 1;
    for (int i = n - 1; i > 0; i--)
    {
        if (A[i] > A[i - 1])
        {
            pi = i - 1;
            desc = 0;
            break;
        }
    }
    if (desc)
    {
        reverse(A.begin(), A.end());
        return;
    }
    // db1(pi);
    for (int i = n - 1; i > pi; i--)
    {
        if (A[i] > A[pi])
        {
            pj = i;
            break;
        }
    }
    swap(A[pi], A[pj]);
    // rep(i, n) cout << A[i] << " ";
    reverse(A.begin() + pi + 1, A.end());
}
int main()
{
    FastIO
        ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;
        vector<int> v(n);
        rep(i, n) cin >> v[i];
        next_permutation(v);
        rep(i, n) cout << v[i] << " ";
        cout << '\n';
    }
    return 0;
}

File: nextPermutation.py
from os import *
from sys import *
from collections import *
from math import *


def nextPermutation(permutation, n) -> list[int]:
    """
    This function takes a permutation and its length as input, and returns the next permutation in lexicographical order.
    
    The approach is to find the first decreasing pair from the right, and swap the first of the pair with the smallest element greater than it from the right.
    
    Finally, reverse the permutation from the position of the first of the pair to the end.
    """
    
    i = 0
    if n > 1:
        i = n - 2
    else:
        return permutation
    point1 = -1
    point2 = -1
    while i >= 0:
        if permutation[i] < permutation[i + 1]:
            point1 = i
            break
        i -= 1
    if point1 == -1:
        return sorted(permutation)
    i = n - 1
    while i > point1:
        if permutation[i] > permutation[point1]:
            point2 = i
            break
        i -= 1
    permutation[point1], permutation[point2] = permutation[point2], permutation[point1]
    permutation[point1 + 1 : n] = reversed(permutation[point1 + 1 : n])
    return permutation


arr = [1, 2, 3, 4]
print(nextPermutation(arr, 4))
